[geeliette] iP

geeliette opened the pull request:
# Fleur: Your Key To Blooming 🌸 
> "A good system shortens the road to the goal." - Orison Swett Marden

**Fleur** is the system for you to keep track of all your daily tasks, so that you can achieve your goals more efficiently. 

## Why You'll Love Fleur
- <ins>Text-based</ins> 📄 : Simple and straightforward interaction
- <ins>User-friendly</ins> 🩵 : Intuitive commands that anyone can learn
- <ins>Lightning fast</ins> ⚡ : Instant feedback and speedy task management

## How To Use It
1. Download Fleur from [here](https://github.com/geeliette/ip/releases/download/A-Release/fleur.jar)
2. Double-click it or run it using the command: `java -jar "fleur.jar"`
3. Start adding your tasks
4. Sit back and enjoy a more organised life! 🦔 

And it's ***FREE***!

## Features
- [x] Manage tasks (ToDos, Deadlines and Events)
- [x] Find tasks using keywords
- [ ] Reminders (coming soon)

## Main Method
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        new Fleur("fleur.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

geeliette committed:
Change name and greeting

geeliette committed:
Update Fleur.java

geeliette committed:
Echo user

geeliette committed:
Add task storing

geeliette committed:
Add Task class

geeliette committed:
Add task types

geeliette committed:
Test using I/O

geeliette committed:
Add Exceptions

geeliette committed:
Add delete function

geeliette committed:
Add save to hard disk function

geeliette committed:
Merge branch 'branch-Level-7'

geeliette committed:
Change dates to LocalDate

geeliette committed:
Merge branch 'branch-Level-8'

geeliette committed:
Add new Ui, Storage, Parser and TaskList classes

geeliette committed:
Use more OOP

geeliette committed:
Merge branch 'branch-A-MoreOOP'

geeliette committed:
Fix bugs

charlenelau-20 commented on src/main/java/fleur/Parser.java:
```diff
@@ -0,0 +1,115 @@
+
+    public void parse(String str) {
+        try {
+            if (str.equalsIgnoreCase("bye")) {
+                exit();
+            } else if (str.equalsIgnoreCase("list")) {
+                this.taskList.listTasks();
+            } else if (str.startsWith("mark")) {
+                int index = Integer.parseInt(str.substring(5)) - 1;
+                this.taskList.markDone(index);
+                System.out.println("Enchanté! I 'ave marked zis task as done:");
+                System.out.println(this.taskList.getTask(index).toString());
+            } else if (str.startsWith("unmark")) {
+                int index = Integer.parseInt(str.substring(7)) - 1;
+                this.taskList.markUndone(index);
+                System.out.println("Zut! I 'ave marked zis task as not done:");
+                System.out.println(this.taskList.getTask(index).toString());
+            } else if (str.startsWith("todo")) {
+                addToDo(str);
+            } else if (str.startsWith("deadline")) {
+                addDeadline(str);
+            } else if (str.startsWith("event")) {
+                addEvent(str);
+            } else if (str.startsWith("delete")) {
+                int index = Integer.parseInt(str.substring(7)) - 1;
+                Task deletedTask = this.taskList.getTask(index);
+                System.out.println("D'accord, I 'ave removed zis task from your list:");
+                System.out.println(deletedTask.toString());
+                this.taskList.deleteTask(index);
+                System.out.println("Now you 'ave " + this.taskList.size() + " task(s) in your list.");
+            } else {
+                throw new FleurInvalidCommandException();
+            }
```
Maybe using switch case statements could make the code easier to read?

charlenelau-20 commented on src/main/java/fleur/Storage.java:
```diff
@@ -0,0 +1,79 @@
+        String str;
+        while ((str = br.readLine()) != null) {
+            Task t = null;
+            switch (str.charAt(1)) {
+                case 'T':
+                    t = new ToDo(str.substring(7));
+                    break;
+                case 'D':
+                    String desc = str.substring(7).split("\\(by: ")[0];
+                    String date = str.substring(7).split("\\(by: ")[1].replace(")", "");
+                    LocalDate by = LocalDate.parse(date, INPUT);
+                    t = new Deadline(desc, by);
+                    break;
+                case 'E':
+                    String[] arr = str.substring(7).split("\\(from: ");
+                    String description = arr[0];
+                    String from = arr[1].split("to: ")[0].trim();
+                    String to = arr[1].split("to: ")[1].replace(")", "");
+                    LocalDate dateFrom = LocalDate.parse(from, INPUT);
+                    LocalDate dateTo = LocalDate.parse(to, INPUT);
+                    t = new Event(description, dateFrom, dateTo);
+                    break;
```
Just a minor layout fix; remember not to have indentations for case clauses in this course!

kai824 replied:
Remember to configure your IDE not to have indentations for case clauses for this course

charlenelau-20 commented on src/main/java/fleur/Ui.java:
```diff
@@ -0,0 +1,22 @@
+        this.scanner = new Scanner(System.in);
+    }
+
+    public void welcomeMessage() {
```
Perhaps using a verb for the method name, e.g. showWelcomeMessage(), could make it more intuitive.

charlenelau-20 commented on src/main/java/fleur/Storage.java:
```diff
@@ -0,0 +1,79 @@
+                    t = new ToDo(str.substring(7));
+                    break;
+                case 'D':
+                    String desc = str.substring(7).split("\\(by: ")[0];
+                    String date = str.substring(7).split("\\(by: ")[1].replace(")", "");
+                    LocalDate by = LocalDate.parse(date, INPUT);
+                    t = new Deadline(desc, by);
+                    break;
+                case 'E':
+                    String[] arr = str.substring(7).split("\\(from: ");
```
Maybe using full/more descriptive words for the variable names 'desc' and 'arr' would make it easier for readers to understand what they are used for.

charlenelau-20 reviewed (commented):
Looks good overall! Also liked how unique the responses are!

geeliette committed:
Fix bugs in testing

geeliette committed:
Merge branch 'add-gradle-support'

kai824 commented on src/main/java/fleur/Storage.java:
```diff
@@ -0,0 +1,79 @@
+        this.dataFile = dataFile;
+    }
+
+    protected ArrayList<Task> loadData() throws IOException, FleurCorruptFileException {
```
> `protected`

Why do you choose to use protected here? Will there be any class that will inherit from Storage and need access to this method?

kai824 commented on src/main/java/fleur/Fleur.java:
```diff
@@ -0,0 +1,181 @@
+//    private void addEvent(String str) throws FleurMissingDetailsException, FleurInvalidDateException {
+//        try {
+//            String[] arr = str.substring(6).split("/from");
+//            String desc = arr[0];
+//            String from = arr[1].split("/to")[0].trim();
+//            String to = arr[1].split("/to")[1].trim();
+//            LocalDate dateFrom = LocalDate.parse(from, INPUT);
+//            LocalDate dateTo = LocalDate.parse(to, INPUT);
+//            Task t = new Event(desc, dateFrom, dateTo);
+//            tasks.add(t);
+//            count++;
+//            System.out.println("Bah, oui! I 'ave added zis event task to your list:");
+//            System.out.println(t.toString());
+//            System.out.println("Now you 'ave " + count + " task(s) in your list.");
+//        } catch (ArrayIndexOutOfBoundsException | StringIndexOutOfBoundsException e) {
+//            throw new FleurMissingDetailsException();
+//        } catch (DateTimeParseException e) {
+//            throw new FleurInvalidDateException();
+//        }
+//    }
+//
+//    private void deleteTask(int index) {
+//        Task deletedTask = tasks.get(index);
+//        count--;
+//        System.out.println("D'accord, I 'ave removed zis task from your list:");
+//        System.out.println(deletedTask.toString());
+//        System.out.println("Now you 'ave " + count + " task(s) in your list.");
+//        tasks.remove(index);
+//    }
+//
+//    private void exit() {
+//        try {
+//            saveData();
+//            System.out.println("Au revoir, 'ope to see you again soon!");
+//        } catch (IOException e) {
+//            System.out.println("Mince! Zere was an error when saving your tasks.");
+//        }
+//    }
+//
+//}
```
Having such a block of commented code may not be the best practice. Consider deleting it

kai824 commented on src/main/java/fleur/Parser.java:
```diff
@@ -0,0 +1,115 @@
+            LocalDate dateFrom = LocalDate.parse(from, INPUT);
+            LocalDate dateTo = LocalDate.parse(to, INPUT);
+            Task t = new Event(desc, dateFrom, dateTo);
+            this.taskList.addTask(t);
+            System.out.println("Bah, oui! I 'ave added zis event task to your list:");
+            System.out.println(t.toString());
```
While this Task object is temporary, consider using a more descriptive name (e.g. `taskToAdd`)

kai824 commented on src/main/java/fleur/Parser.java:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    private void addEvent(String str) throws FleurMissingDetailsException, FleurInvalidDateException {
```
Good job specifying the two checked Exceptions separately instead of just specifying `throws FleurException`, so that the callers can choose to handle them separately.

kai824 commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,13 @@
+list
+unmark 2
+list
+delete 3
+hi
+todo
```
Good job testing the edge cases and error messages! Consider testing the case of index out of bounds as well

kai824 commented on src/main/java/fleur/FleurException.java:
```diff
@@ -0,0 +1,7 @@
+package fleur;
+
+public class FleurException extends Exception {
```
Perhaps you could group the Exception classes together into 1 package to make the file organisation even tidier

kai824 reviewed (commented):
Looks good overall! Love the creativity in the use of French in your responses.

Do remember to add javadoc comments.

geeliette committed:
Add gradle

geeliette committed:
Add JUnit tests

geeliette committed:
Add JAR file

geeliette committed:
Edit Fleur class

geeliette committed:
Add JavaDoc comments

geeliette committed:
Match Java coding standard

geeliette committed:
Add find function

geeliette committed:
Merge branch 'A-JavaDoc'

geeliette committed:
Fix merge conflicts

geeliette committed:
Merge branch 'branch-Level-9'

geeliette committed:
Fix naming error

geeliette committed:
Add packages

geeliette committed:
Fix bugs

geeliette committed:
Add Commands class

geeliette committed:
Fix exception handling

geeliette committed:
Merge branch 'branch-A-Commands'

geeliette committed:
Add GUI

geeliette committed:
Merge branch 'branch-Level-10'

geeliette committed:
Add assert statements to enforce key assumptions in input validation

The current implementation does not enforce certain key assumptions
about user input and task modifications. Task indices in commands like
delete, mark, and unmark should be more than 0 and less than the total
number of tasks. Additionally, parse() does not explicitly check for
empty input strings before processing them.

These missing validations can lead to unexpected runtime errors,
causing the program to crash instead of failing early with a clear
message.

Let's add assert statements in critical parts of the code to ensure
valid preconditions hold before execution. Specifically:
* Add assertions in parse() to validate that input is non-null and
  non-empty
* Introduce assertions in execute() method in DeleteCommand,
  MarkCommand, and UnmarkCommand to check that task indices are within
  a valid range

geeliette committed:
Refactor Storage and Command classes to improve task loading and saving

The loadData() method in Storage is too long and tightly coupled
with the task creation logic. It attempts to parse raw text data,
determine task types, and instantiate the correct task objects within
a single method.

The method contains more details than what the reader can process at
a time, making it difficult to maintain and extend in the future.

Refactoring loadData() by adding new methods in the Command classes
allows us to abstract and move task creation logic to a more appropriate
place, improving code quality and readability.

Let's,
* introduce createToDo(), createDeadline(), and createEvent() in
  AddToDoCommand, AddDeadlineCommand, and AddEventCommand
* modify loadData() to delegate task creation to these methods instead
  of handling it directly

geeliette committed:
Merge pull request #2 from geeliette/branch-A-Assertions

Add assert statements to enforce key assumptions in input validation

geeliette committed:
Merge pull request #3 from geeliette/branch-A-CodeQuality

Refactor Storage and Command classes to improve task loading and saving

geeliette committed:
Add edit command class

geeliette committed:
Shorten edit execute method

geeliette committed:
Merge branch 'branch-C-Update'

geeliette committed:
Update GUI

geeliette committed:
Merge branch 'branch-A-BetterGui'

geeliette committed:
Display welcome message

geeliette committed:
Add representative screenshot

geeliette committed:
Update README.md
